qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+            and sig1.dt == sig2.dt
+            and len(sig1.samples) == len(sig2.samples)
+        ):
+            pwc1 = DiscreteSignal(
+                dt=sig2.dt,
+                samples=0.5 * sig1.samples * sig2.samples,
+                start_time=sig2.start_time,
+                duration=sig2.duration,
+                carrier_freq=sig1.carrier_freq + sig2.carrier_freq,
+                phase=sig1.phase + sig2.phase,
+            )
+            pwc2 = DiscreteSignal(
+                dt=sig2.dt,
+                samples=0.5 * sig1.samples * np.conjugate(sig2.samples),
+                start_time=sig2.start_time,
+                duration=sig2.duration,
+                carrier_freq=sig1.carrier_freq - sig2.carrier_freq,
+                phase=sig1.phase - sig2.phase,
+            )
+            return pwc1 + pwc2
+
+    # if no special cases apply, implement generic rule
+    def new_env1(t):
+        return 0.5 * sig1.envelope(t) * sig2.envelope(t)
+
+    def new_env2(t):
+        return 0.5 * sig1.envelope(t) * np.conjugate(sig2.envelope(t))
+
+    prod1 = Signal(
+        envelope=new_env1,
+        carrier_freq=sig1.carrier_freq + sig2.carrier_freq,
+        phase=sig1.phase + sig2.phase,
+    )
+    prod2 = Signal(
+        envelope=new_env2,
+        carrier_freq=sig1.carrier_freq - sig2.carrier_freq,
+        phase=sig1.phase - sig2.phase,
+    )
+    return prod1 + prod2
+
+
+def sort_signals(sig1: Signal, sig2: Signal) -> Tuple[Signal, Signal]:
+    """Utility function for ordering a pair of ``Signal``s according to the partial order:
+    ``sig1 <= sig2`` if and only if:
+        - ``type(sig1)`` precedes ``type(sig2)`` in the list
+        ``[Constant, DiscreteSignal, Signal, SignalSum, DiscreteSignalSum]``.
+    """
+    if isinstance(sig1, Constant):
+        pass
+    elif isinstance(sig2, Constant):
+        sig1, sig2 = sig2, sig1
+    elif isinstance(sig1, DiscreteSignal):
+        pass
+    elif isinstance(sig2, DiscreteSignal):
+        sig2, sig1 = sig1, sig2
+    elif isinstance(sig1, Signal):
+        pass
+    elif isinstance(sig2, Signal):
+        sig2, sig1 = sig1, sig2
+    elif isinstance(sig1, SignalSum):
+        pass
+    elif isinstance(sig2, SignalSum):
+        sig2, sig1 = sig1, sig2
+
+    return sig1, sig2
+
+
+def to_SignalSum(sig: Union[int, float, complex, Array, Signal]) -> SignalSum:
+    """Convert the input to a SignalSum according to:
+
+    - If it is already a SignalSum, do nothing.
+    - If it is a Signal but not a SignalSum, wrap in a SignalSum.
+    - If it is a number, wrap in Constant in a SignalSum.
+    - Otherwise, raise an error.
+
+    Args:
+        sig: A SignalSum compatible input.
+
+    Returns:
+        SignalSum
+
+    Raises:
+        QiskitError: If the input type is incompatible with SignalSum.
+    """
+
+    if isinstance(sig, (int, float, complex)) or (isinstance(sig, Array) and sig.ndim == 0):
+        sig = Constant(sig)
```

Comment:
Is `Constant` a `SignalSum`? Does this work with the error raised below?

Reply:
The subsequent check wraps any subclass of `Signal` that is not a `SignalSum` in a `SignalSum`, so this will handle the `Constant` case.